Add climate group

daenny opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds a climate group component. Based initially on https://github.com/daenny/climate_group, but updated to be similar to the existing light/fan implementations.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/23990

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hello @daenny,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`ae5b3c43fb3ed250a8b50490ec3edad764033590`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;77737&#x2F;commits&#x2F;ae5b3c43fb3ed250a8b50490ec3edad764033590) - This commit has something that looks like an email address (git@danielclaes.de). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

daenny committed:
feat: initial climate group for hassio

boswelja commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,356 @@
+    HVACAction.FAN,
+    HVACAction.IDLE,
+    HVACAction.OFF,
+    None,
```
What does `None` denote here?

daenny replied:
Indeed not needed anymore. I previously needed it to also be able to handle devices without HVAC action, but the new way defaults to None in any case

boswelja commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,356 @@
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Initialize Light Group config entry."""
```
This comment is not accurate, I don't think we're setting up light groups here :)

daenny replied:
Adjusted!

boswelja commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,356 @@
+        self._attr_unique_id = unique_id
+        self._attr_extra_state_attributes = {ATTR_ENTITY_ID: entity_ids}
+
+        if "c" in unit.lower():
```
Can we pass `TEMP_CELSIUS` and `TEMP_FAHRENHEIT` in directly? That way we only have a single-line assignment

boswelja commented on homeassistant/components/group/climate.py:
```diff
@@ -0,0 +1,356 @@
+        # Set group as unavailable if all members are unavailable or missing
+        self._attr_available = any(state.state != STATE_UNAVAILABLE for state in states)
+
+        def _mean(*args: float) -> float:
```
Can we use `statistics.mean` here?
https://docs.python.org/3/library/statistics.html#statistics.mean

daenny replied:
Done!

daenny committed:
chore: adjusted to review comments

daenny commented:
Thanks for the review @boswelja , I adjusted the code accordingly!

daenny commented:
Bump @balloob or @frenck , I saw that you guys merged some previous PRs for the group integration. Could you please have a look at it?

ignisf commented:
@daenny, thank you for contributing this, I am eagerly awaiting its merge. Maybe the pylint warning needs to be looked into as well?

daenny commented:
Yes, that's a new one after making my branch upstream.  I will look at it tomorrow

KartoffelToby commented:
@daenny  I wonder how the local calibration entities and also the valve_position entities can interact, or is it possible to query the values directly within the group via the attributes of the climate devices? since they are separate entities this is confusing. the climate entity still has the attributes local_temperature_calibration and valve_position, is it still possible to change them directly, or do you have to go through your own entities?

KartoffelToby commented:
Many climate Intigrations have different Services to set a Offset/calibration think its time to bundle them under a global Services like service climate set offset

daenny committed:
fix: pylint

Quentame reviewed (commented):
At first glace, it looks good to me !

Will test with my devices (which have some non core presets, see #72790) + check some group integration dehaviour before merging.

Quentame reviewed (changes requested):
The overall code is working, even custom presets are handled, but:
- you should add the climate group to the `config_flow`, so we can configure it into the UI. Otherwise `notify` and `climate` will be the only groups to not be able to be configured by the UI.
- the climate group can't be added into a climate card  --> frontend issue I think
<img width="999" alt="climate-card-climate-group" src="https://user-images.githubusercontent.com/14821482/201675562-a1e365d9-090d-407a-8c62-34b8022849f6.png">


The following concerns the `overkiz` climates I am using, as it did not appear all the time and having a log on the `overkiz` side, but still quite often to be noticed (and may still be catch/handled properly/differently ?):
- when setting a temperature/mode/preset: it sets it well, but then it shows a `climate/set_hvac_mode` toast error ->  make the group unavailable (for a sec or less) -> make the group available but without the current mode & preset. Exiting the popup and clicking again fill the current mode & preset again. The current target temperature is displayed correctly.

Having those logs during the process
```shell
2022-11-14 14:20:02.023 ERROR (MainThread) [homeassistant.components.overkiz] Error fetching device events data: Too many requests, try again later.
2022-11-14 14:20:03.024 INFO (MainThread) [homeassistant.components.overkiz] Fetching device events data recovered
```

I think there are 3 issues in here:
1. the _Too many requests_ error log from `overkiz` (the root issue)
2. the `climate/set_hvac_mode` toast error, certainly caused by the first issue
3. the climate popup not displaying current mode & preset (lists are working) --> frontend issue ? (labels are on top, like some value were filled) OR group climate values are `None` ?

Screenshot of the process:
1. setting a preset
<img width="1120" alt="screenshot-preset-1" src="https://user-images.githubusercontent.com/14821482/201672767-a0e29303-b098-4677-97f2-d9ab97c3d8fb.png">

2. the group is unavailable
<img width="1120" alt="screenshot-preset-2" src="https://user-images.githubusercontent.com/14821482/201672825-4b0dd86b-5416-4a29-8c55-03d5f4bde1a7.png">

3. the group is back but current mode & preset are not filled
<img width="1120" alt="screenshot-preset-3" src="https://user-images.githubusercontent.com/14821482/201672856-f8ca4ae8-b44b-4364-a543-ea28f543e0c4.png">

CC @iMicknl & @tetienne from the `overkiz` integration.

---
Translations
"thermostat" -> climate
"température cible" -> target temperature
"chauffe" -> "heating" HVACMode.HEAT
"actuellement" -> now

KartoffelToby commented:
@daenny i have a custom Integration that have a own climate group function (i need a little Special here, because i need to perform calibration for each climate) but i think it can Help you. Because i have issues too to restore all values at Startup. I ended up with a "wait for all climate" Routine. Think its not the most elegant solution, but it works.

Little side note, i wounder if all "custom" Attributes from the climates are pass throw and merged

daenny committed:
fix: return most common mode

daenny commented:
@KartoffelToby thanks if you are happy to help! if you have a code snippet or link please let me know! 
I just pushed changes that the most common mode is reported instead of the priority. That's also more in line with the other attributes. 
@Quentame thanks for testing and the review. I have not added things to config_flow before. Is there a good reference?

daenny committed:
fix: added config flow

KartoffelToby commented:
@daenny for config flow: https://github.com/KartoffelToby/better_thermostat/blob/e296464145b6d6867d9178ed0ebc78cef3e13092/custom_components/better_thermostat/config_flow.py#L211

for the startup: https://github.com/KartoffelToby/better_thermostat/blob/e296464145b6d6867d9178ed0ebc78cef3e13092/custom_components/better_thermostat/climate.py#L359

but i dont know if its the right way to restore or get values on startup. feel free to ask me if you need help, i worked a lot with climates stuff as you can see in this custom interation better thermostat

daenny committed:
fix: black and flake8

Quentame commented:
> @Quentame thanks for testing and the review. I have not added things to config_flow before. Is there a good reference?

You actually don't need to create a config flow from the scratch here, there is already a `config_flow.py` in the `group` integration, take the exemple of the `light` platform in the file.

Also, you will need to update the `strings.json` file accordingly, and launch command to create the translation files.

And, to finish (I promiss), you should add tests from your config flow additions.

---

Links to current files:

`config_flow.py`
https://github.com/home-assistant/core/blob/dev/homeassistant/components/group/config_flow.py

`strings.json`
https://github.com/home-assistant/core/blob/dev/homeassistant/components/group/strings.json

`test_config_flow.py`
https://github.com/home-assistant/core/blob/dev/tests/components/group/test_config_flow.py

Quentame commented:
Seems to work better but still having the same issue sometimes (I think less)

<img width="567" alt="image" src="https://user-images.githubusercontent.com/14821482/201740227-9aae7ba6-cb8b-4b98-9b45-68e090c6679b.png">

iMicknl commented:
@Quentame could you perhaps log this in a separate issue? With some information about which devices + how many in the group. I assume this is due to the fact that we send multiple commands per device, and when you group multiple devices, it will send too many commands.

We are still looking into combining multiple commands in one execution, but this work hasn't been prioritized yet.

Quentame commented:
@daenny : Can you fix tests + add test of your config flow  + add translation ?

translation: on the `strings.json` file as described here https://github.com/home-assistant/core/pull/77737#issuecomment-1314073937 (I don't remember the command line to run after updating the string.json file, but otherwise, fill the en.json manually)

rjulius23 commented:
I think it should be possible to choose different aggregations, right now only mean can be used, and if I understand correctly it calculates the mean of the climate entities in the group for the current temperature, but if ANY 1 of them is in heating mode the groups entity will show heating, which can be misleading, so having hte possibility to use the min value would be more accurate IMHO.

ignisf commented:
Please do not close this PR

Is there anything from @Quentame feedback left unaddressed?

boswelja commented:
Seems like tests were still failing, though logs aren't available anymore

daenny committed:
test: added strings and test_config flow

daenny commented:
Sorry, I was a bit distracted the last couple of months for personal reasons. I also lost a bit motivation, because I do not use the component myself anymore. But anyway, I made the PR upstream and addressed the changes from @Quentame for the best of my knowledge. 
I will try to get this one merged now, so let me know if I am missing something.

daenny committed:
fix: use correct attributes

daenny committed:
fix: typo

daenny committed:
fix: mypy

daenny committed:
fix: map translations string to binary sensor

daenny commented:
Okay, it seems something has changed in the meantime with the climate demo group. I will try to fix those tomorrow.

daenny committed:
fix: tests with climate group

Quentame commented:
Hi everyone !

Yes there is still some leftovers comments in this PR as I remember (config_flow + strings for translation).

Forget about my issues I got from the `overkiz` integration.

Thanks 💪

daenny commented:
> Hi everyone !
> 
> Yes there is still some leftovers comments in this PR as I remember (config_flow + strings for translation).
> 
> Forget about my issues I got from the `overkiz` integration.
> 
> Thanks 💪

I updated the config flow and `strings.json` in the latest commits and added the tests. For the tests l, I needed to change the state reporting in the `ClimateEntity`. Can you check if everything is as you expect?
Thanks!

Devil-of-Chaos commented:
Is there any update on this?

KartoffelToby commented:
Think its a bigger Problem that most users think, i have create a climate group function in my own Integration https://github.com/KartoffelToby/better_thermostat and i saw a lot of possible Problems, sync of device entities Like valve, calibration, etc. That shoudn't be synced between the group members.

ignisf commented:
Just chiming in to mostly tell the bot off and that it would be awesome to have this for the heating season <3

magicmellon commented:
> Just chiming in to mostly tell the stale bot off and that it would be awesome to have this for the heating season <3

Totally agree - started using this addon today and it made hours of node-red making take an hour of refactoring. Would be awesome to see an official implimentation as it is one of the few things that isnt UI Configurable as a group.

vordenken commented:
Would love to see some progress on this as well. Can't help but would love to test if anything needs it.

edenhaus reviewed (changes requested):
Can you please rebase and fix the merge conflicts?

After that, I can start reviewing your PR

r0bb10 commented:
@daenny any news on the fixes? so it could be reviewed and maybe finally added in core.

andrewthauer commented:
I found myself wanting this feature and stumped across this PR. It seems it's based on https://github.com/daenny/climate_group which suggests this PR may be abandoned.

r0bb10 commented:
> I found myself wanting this feature and stumped across this PR. It seems it's based on https://github.com/daenny/climate_group which suggests this PR may be abandoned.

yeah i saw that too, this wont probably be addressed until someone expert enough finishes forks and pr again.

jayme-github commented:
> Can you please rebase and fix the merge conflicts?
> 
> After that, I can start reviewing your PR

Hey, I did a rebase with small adaptions, updated strings.json and did local testing at https://github.com/jayme-github/home-assistant-core/tree/feature/climate-group. As I can probably not re-use this PR, should I create a new one for you to review? I'm up to take care of follow ups to get this merged.

r0bb10 commented:
> > Can you please rebase and fix the merge conflicts?
> > After that, I can start reviewing your PR
> 
> Hey, I did a rebase with small adaptions, updated strings.json and did local testing at https://github.com/jayme-github/home-assistant-core/tree/feature/climate-group. As I can probably not re-use this PR, should I create a new one for you to review? I'm up to take care of follow ups to get this merged.

oh wow this is really a good news! keep it up!

ignisf commented:
> should I create a new one for you to review?

Could you please leave a reference to the new PR here so we can track it?

jascdk commented:
any thoughts of if this is gonna be merged into HA?

r0bb10 commented:
> should I create a new one for you to review? I'm up to take care of follow ups to get this merged.

@jayme-github any news?

albertomontesg commented:
Trying to revamp this work with a parallel PR which I would be happy to follow up and make sure it is being merged: https://github.com/home-assistant/core/pull/114290
